solidity/contracts/isms/hook/layer-zero/LayerZeroV2Ism.sol

```diff
@@ -0,0 +1,134 @@
+  @@@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+@@@@@@@@@       @@@@@@@@*/
+
+// ============ Internal Imports ============
+
+import {IInterchainSecurityModule} from "../../../interfaces/IInterchainSecurityModule.sol";
+import {Message} from "../../../libs/Message.sol";
+import {TypeCasts} from "../../../libs/TypeCasts.sol";
+import {AbstractMessageIdAuthorizedIsm} from "../AbstractMessageIdAuthorizedIsm.sol";
+
+// ============ External Imports ============
+import {Origin} from "@layerzerolabs/lz-evm-protocol-v2/contracts/interfaces/ILayerZeroEndpointV2.sol";
+
+/**
+ * @title LayerZeroV2Ism
+ * @notice Uses LayerZero V2 deliver and verify a messages Id
+ */
+contract LayerZeroV2Ism is AbstractMessageIdAuthorizedIsm {
+    using Message for bytes;
+    using TypeCasts for bytes32;
+
+    // Layerzero endpoint address
+    address public immutable endpoint;
+
+    // ============ Constants ============
+
+    uint8 public constant moduleType =
+        uint8(IInterchainSecurityModule.Types.NULL);
+
+    // @dev the offset of msg.data where the function parameters (as bytes) begins. 4 bytes is always used when encoding the function selector
+    uint8 constant FUNC_SELECTOR_OFFSET = 4;
+
+    // ============ Constructor ============
+    constructor(address _endpoint) {
+        require(
+            _endpoint != address(0),
+            "LayerZeroV2Ism: invalid authorized endpoint"
+        );
+        endpoint = _endpoint;
+    }
+
+    /**
+     * @notice Entry point for receiving msg/packet from the LayerZero endpoint.
+     * @param _message The payload of the received message.
+     * @dev Authorization verifcation is done within verifyMessageId() -> _isAuthorized()
+     */
+    function lzReceive(
+        Origin calldata,
+        bytes32,
+        bytes calldata _message,
+        address,
+        bytes calldata
+    ) external payable {
+        verifyMessageId(_messageId(_message));
+    }
+
+    // ============ Internal function ============
+
+    /**
+     * @notice Slices the messageId from the message delivered from LayerZeroV2Hook
+     * @dev message is created as abi.encodeCall(AbstractMessageIdAuthorizedIsm.verifyMessageId, id)
+     * @dev _message will be 36 bytes (4 bytes for function selector, and 32 bytes for messageId)
+     */
+    function _messageId(
+        bytes calldata _message
+    ) internal pure returns (bytes32) {
+        return bytes32(_message[FUNC_SELECTOR_OFFSET:]);
+    }
+
+    /**
+     * @notice Validates criterias to verify a message
+     * @dev this is called by AbstractMessageIdAuthorizedIsm.verifyMessageId
+     * @dev parses msg.value to get parameters from lzReceive()
+     */
+    function _isAuthorized() internal view override returns (bool) {
+        require(_isAuthorizedHook(), "LayerZeroV2Ism: hook is not authorized");
+
+        require(
+            _isVerifyMessageSelector(),
+            "LayerZeroV2Ism: message payload is incorrect"
+        );
```

Comment:
idk if this is totally necessary

Reply:
we are already authenticating the origin's sender, which will only send this type of message

Reply:
true. Removed `_isVerifyMessageSelector()`